### ISSUE TITLE
Flake eval: --show-trace should imply eval-cache false

### ISSUE THREAD

**Issue by picnoir** [NixOS/nix#9165]
**Describe the bug**

If a flake (cached) evaluation fails, a subsequent evaluation with `--show-trace` won't display the full error trace but an “error: cached failure of attribute” unhelpful message.

The current workaround consists in setting `--option eval-cache false` or adding a new line in the flake definition to invalidate the cache.

**Steps To Reproduce**

```
/etc/nixos (master*) » nixos-rebuild build --flake .#trantor                                                        
building the system configuration...
warning: Git tree '/etc/nixos' is dirty
error:
       … while calling the 'seq' builtin

         at /nix/store/zwq17yl8f5nvsy5iim2vd4svw82ci9nv-source/lib/modules.nix:320:18:

          319|         options = checked options;
          320|         config = checked (removeAttrs config [ "_module" ]);
             |                  ^
          321|         _module = checked (config._module);

       … while evaluating a branch condition

         at /nix/store/zwq17yl8f5nvsy5iim2vd4svw82ci9nv-source/lib/modules.nix:261:9:

          260|       checkUnmatched =
          261|         if config._module.check && config._module.freeformType == null && merged.unmatchedDefns != [] then
             |         ^
          262|           let

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: cannot look up '<nixpkgs/nixos/modules/installer/scan/not-detected.nix>' in pure evaluation mode (use '--impure' to override)

       at «none»:0: (source not available)

/etc/nixos (master*) » nixos-rebuild build --flake .#trantor --show-trace                                                                                                                                                          
building the system configuration...
warning: Git tree '/etc/nixos' is dirty
error: cached failure of attribute 'nixosConfigurations.trantor.config'
```

**Expected behavior**

When `--show-trace` is set, the full trace should be printed instead of cache failure error message.

**`nix-env --version` output**

2.17.0

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by SuperSandro2000** [comment]
I think the problem at hand here is rather that failed evaluations are cached, not that --show-trace should forcefully disable the evaluation cache.

**Comment by picnoir** [comment]
Does Hydra or some other large scale evaluator leverages the cache failures at the moment?

If it does, I guess there's some value in caching the failures.

**Comment by quinn-dougherty** [comment]
:heavy_plus_sign:

**Comment by SuperSandro2000** [comment]
> If it does, I guess there's some value in caching the failures.

Then they should opt into this via some config or argument but for cli users this is rather annoying and confusing.

**Comment by eclairevoyant** [comment]
> I think the problem at hand here is rather that failed evaluations are cached, not that --show-trace should forcefully disable the evaluation cache.

Eelco is against caching error messages, but wants the failure itself to be cached. So we end up back here where `--show-trace` has to implicitly disable the eval cache.

**Comment by Dummyc0m** [comment]
What's the use case for caching errors? Because for the end user, I had to end up on this bug.

**Comment by eclairevoyant** [comment]
Please follow https://github.com/NixOS/nix/pull/10564

**Comment by roberth** [comment]
Fixed in #10564
Specifically, it makes errors cause re-evaluation, regardless of whether `show-trace` is enabled or not.
